### PR TITLE
improve compression size estimation in shadow compression

### PR DESCRIPTION
**Pull request by roberto-bayardo**
**Description**

Shadow compression currently executes one Flush() per Write(), which can result in an under-estimate of the ultimate compression ratio. This PR implements a heuristic to avoid flushing any data until there is some chance the compressed data size will exceed the target to better pack transactions into frames.

**Tests**

Extended tests to confirm the upperbound compressed size estimate is indeed an upperbound up to frame sizes approaching the max blob size (but not beyond it).

Have it running on base-goerli devnet, and so far so good.

**Additional context**

Confirmed that this new scheme can result in more transactions packed into a single frame than previous, though have yet to do an analysis on transactions in the wild.

### CODE REVIEW

File: op-batcher/compressor/shadow_compressor.go [hunk 1]
```diff
@@ -41,26 +54,39 @@ func NewShadowCompressor(config Config) (derive.Compressor, error) {
...
+	if newBound > cap {
+		// Do not flush the buffer unless there's some chance we will be over the size limit.
+		// This reduces CPU but more importantly it makes the shadow compression ratio more
+		// closely reflect the ultimate compression ratio.
+		err = t.shadowCompress.Flush()
+		if err != nil {
+			return 0, err
+		}
+		newBound = uint64(t.shadowBuf.Len()) + 4 // + 4 is to account for the digest written on close()
+		if newBound > t.config.TargetFrameSize*uint64(t.config.TargetNumFrames) {
```
Comment (mdehoog) [inline op-batcher/compressor/shadow_compressor.go]: nit: could use `cap` here?

File: op-batcher/compressor/shadow_compressor.go [hunk 2]
```diff
@@ -41,26 +54,39 @@ func NewShadowCompressor(config Config) (derive.Compressor, error) {
...
+	cap := t.config.TargetFrameSize * uint64(t.config.TargetNumFrames)
+	if newBound > cap {
+		// Do not flush the buffer unless there's some chance we will be over the size limit.
+		// This reduces CPU but more importantly it makes the shadow compression ratio more
+		// closely reflect the ultimate compression ratio.
+		err = t.shadowCompress.Flush()
+		if err != nil {
+			return 0, err
+		}
+		newBound = uint64(t.shadowBuf.Len()) + 4 // + 4 is to account for the digest written on close()
```
Comment (mdehoog) [inline op-batcher/compressor/shadow_compressor.go:79]: thanks for the bug fix!

File: op-batcher/compressor/shadow_compressor.go (hunk 1 above)
Comment (roberto-bayardo) [reply op-batcher/compressor/shadow_compressor.go]: yes good catch fixed

File: op-batcher/compressor/shadow_compressor_test.go [hunk 3]
```diff
@@ -110,3 +110,28 @@ func TestShadowCompressor(t *testing.T) {
...
+func TestBoundInaccurateForLargeRandomData(t *testing.T) {
+	var sizeLimit int = 1 << 17
+
+	sc, err := NewShadowCompressor(Config{
+		TargetFrameSize: uint64(sizeLimit + 100),
+		TargetNumFrames: 1,
+	})
+	require.NoError(t, err)
+
+	_, err = sc.Write(randomBytes(t, sizeLimit+1))
```
Comment (sebastianst) [inline op-batcher/compressor/shadow_compressor_test.go:131]: We've recently been moving away from using `crypto/rand` randomness in tests and instead use `math/rand` to make them more deterministic.
I can see that it is practically impossible for this test to fail here given we produce `1 << 17` bytes of randomness, but wondering if we want to take the opportunity to change the randomness in the shadow compressor tests to using a `math/rand.Rand` that we pass to `randomBytes` and that we initialize with a fixed seed in each test.

**Review by sebastianst — APPROVED** [review]
Great improvement, thank you!

Only one suggestion to make randomness in tests deterministic.

File: op-batcher/compressor/shadow_compressor_test.go (hunk 3 above)
Comment (roberto-bayardo) [reply op-batcher/compressor/shadow_compressor_test.go:131]: changed to math/rand

Comment (roberto-bayardo) [reply op-batcher/compressor/shadow_compressor_test.go:131]: ..and with a deterministic generator.

**Review by trianglesphere — APPROVED** [review]

**Pull request merged**